### PR TITLE
Fix array coercion for non-array target type

### PR DESCRIPTION
**Pull request by bajtos**
Before this patch, when sending an array to an argument of type "object" or "any", the value was flattened and the first item was used instead of the array.

This patch fixes the issue by omitting recursive coersion when the target type is not an array.

See #196 that introduced the regression.

/to @ritch please review
/cc @STRML @XinV

### CODE REVIEW

**Comment by STRML** [comment]
Aha. Thanks for the catch. :+1: from me.

File: test/rest.test.js [hunk 1]
```diff
@@ -994,6 +994,37 @@ describe('strong-remoting-rest', function() {
         });
     });
 
+    it('should not flatten arrays for non-standard target type', function(done) {
+      var method = givenSharedStaticMethod(
+        function(arg, cb) { cb(null, { value: arg }); },
+        {
+          // Note: the string "array" is not a valid type
```
Comment (ritch) [inline test/rest.test.js]: why is `array` not a valid type?

Comment (STRML) [inline test/rest.test.js]: Wouldn't the right type be `['any']`?

File: lib/shared-method.js [hunk 2]
```diff
@@ -236,24 +236,25 @@ function escapeRegex(d) {
...
+
+  // If coercing an array to an erray,
+  // then coerce all members of the array too
+  if (targetIsArray && Array.isArray(uarg)) {
+    return uarg.map(function(arg, ix) {
+      // when coercing array items, use only name and type,
+      // ignore all other root settings like "required"
+      return coerceAccepts(arg, {
+        name: name + '[' + ix +']',
+        type: desc.type[0]
```
Comment (ritch) [inline lib/shared-method.js]: this code is very hard to read ...I wonder if we need a `Description` object that has properties like `type` and `arrayValueType` or similar.

File: lib/shared-method.js [hunk 3]
```diff
@@ -236,24 +236,25 @@ function escapeRegex(d) {
...
-    });
-  }
-
   var name = desc.name || desc.arg;
   var targetType = desc.type;
-  var actualType = SharedMethod.getType(uarg);
-
-  // Target type may be an array. If so, expect the arg to be a member of that array,
-  // since we map this function over arrays.
   var targetIsArray = Array.isArray(targetType) && targetType.length === 1;
```
Comment (ritch) [inline lib/shared-method.js]: this should really be named `targetIsTypedArray` or something that makes it clear that we are talking only about `['string']` and similar.

**Comment by ritch** [comment]
Comments can be addressed elsewhere as they aren't really in the scope of this PR.

LGTM

File: lib/shared-method.js (hunk 3 above)
Comment (STRML) [inline lib/shared-method.js]: `targetTypeIsArray` would be good imo - we don't want to say `TypedArray` which is a very different thing.

I agree, we can do cleanup in a separate PR.

File: test/rest.test.js (hunk 1 above)
Comment (bajtos) [inline test/rest.test.js]: See http://docs.strongloop.com/display/LB/LoopBack+types#LoopBacktypes-Arraytypes, these are the valid types:

 - `[String]`
 - `["String"]`
 - `[{type: String, length: 64}]`

In all cases, it's an array with a single item that describes the type of array items.

Comment (bajtos) [inline test/rest.test.js]: For the purpose of the test, I can use `accepts: { arg: 'arg', type: 'any' }` if that's make the test easier to understand.

Comment (bajtos) [inline test/rest.test.js]: See #205.

File: lib/shared-method.js (hunk 3 above)
Comment (bajtos) [inline lib/shared-method.js]: See #205.

File: lib/shared-method.js (hunk 2 above)
Comment (bajtos) [inline lib/shared-method.js]: I agree that such object would be great. However, I don't have time to make this refactoring now :(
